implementation/src/main/java/io/smallrye/mutiny/groups/UniMemoize.java

```diff
@@ -0,0 +1,74 @@
+package io.smallrye.mutiny.groups;
+
+import static io.smallrye.mutiny.helpers.ParameterValidation.nonNull;
+
+import java.time.Duration;
+import java.util.function.BooleanSupplier;
+
+import io.smallrye.common.annotation.Experimental;
+import io.smallrye.mutiny.Uni;
+import io.smallrye.mutiny.infrastructure.Infrastructure;
+import io.smallrye.mutiny.operators.AbstractUni;
+import io.smallrye.mutiny.operators.UniMemoizeOp;
+
+@Experimental("Memoization is an experimental feature at this stage")
+public class UniMemoize<T> {
+
+    private final Uni<T> upstream;
+
+    public UniMemoize(AbstractUni<T> upstream) {
+        this.upstream = nonNull(upstream, "upstream");
+    }
+
+    /**
+     * Memoize the received item or failure as long as the provided boolean supplier evaluates to {@code false}.
+     * <p>
+     * New subscribers will receive the memoize item or failure.
+     * When the boolean supplier evaluates to {@code true} then a new upstream subscription happens and the subscribers get a
+     * chance to observe new values.
+     *
+     * @param invalidationGuard the invalidation guard, which evaluates to {@code false} for as long as the item or failure must
+     *        be memoized
+     * @return a new {@link Uni}
+     */
+    public Uni<T> until(BooleanSupplier invalidationGuard) {
+        return Infrastructure.onUniCreation(new UniMemoizeOp<>(upstream, invalidationGuard));
+    }
+
+    /**
+     * Memoize the received item or failure for a duration after the upstream subscription has been received.
+     * <p>
+     * New subscribers will receive the memoize item or failure.
+     * When duration has elapsed then a new upstream subscription happens and the subscribers get a chance to observe new
```

Comment:
Does it happen at the next subscription attempt after the delay, or when the delay expires?

Reply:
(I got it, just need to be written).